src/clustering/administration/main/command_line.cc

```diff
@@ -645,6 +649,163 @@ service_address_ports_t get_service_address_ports(const std::map<std::string, op
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --web-tls-key and --web-tls-cert");
+        return false;
+    }
+
+    return load_tls_key_and_cert(web_tls, *key_file, *cert_file);
+}
+
+bool configure_driver_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *driver_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--driver-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--driver-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--driver-tls-ca");
+
+    if (!(key_file && cert_file)) {
+        logERR("must specify both --driver-tls-key and --driver-tls-cert");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(driver_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (ca_file) {
+        if (!SSL_CTX_load_verify_locations(driver_tls, ca_file->c_str(), NULL)) {
+            ERR_print_errors_fp(stderr);
+            return false; 
+        }
+
+        // Mutual authentication.
+        SSL_CTX_set_verify(driver_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+    }
+
+    return true;
+}
+
+bool configure_cluster_tls(const std::map<std::string, options::values_t> &opts, SSL_CTX *cluster_tls) {
+    boost::optional<std::string> key_file = get_optional_option(opts, "--cluster-tls-key");
+    boost::optional<std::string> cert_file = get_optional_option(opts, "--cluster-tls-cert");
+    boost::optional<std::string> ca_file = get_optional_option(opts, "--cluster-tls-ca");
+
+    if (!(key_file && cert_file && ca_file)) {
+        logERR("must specify each of --cluster-tls-key, --cluster-tls-cert, and --cluster-tls-ca");
+        return false;
+    }
+
+    if (!load_tls_key_and_cert(cluster_tls, *key_file, *cert_file)) {
+        return false;
+    }
+
+    if (!SSL_CTX_load_verify_locations(cluster_tls, ca_file->c_str(), NULL)) {
+        ERR_print_errors_fp(stderr);
+        return false; 
+    }
+
+    // Mutual authentication.
+    SSL_CTX_set_verify(cluster_tls, SSL_VERIFY_PEER|SSL_VERIFY_FAIL_IF_NO_PEER_CERT, NULL);
+
+    return true;
+}
+
+bool initialize_tls_ctx(SSL_CTX **tls_ctx) {
+    *tls_ctx = SSL_CTX_new(SSLv23_method());
+    if (NULL == *tls_ctx) {
+        ERR_print_errors_fp(stderr);
+        return false;
+    }
+
+    // Only allow TLS v1.2 and prefer server ciphers.
+    SSL_CTX_set_options(*tls_ctx, SSL_OP_NO_SSLv2|SSL_OP_NO_SSLv3|SSL_OP_NO_TLSv1|SSL_OP_NO_TLSv1_1|SSL_OP_CIPHER_SERVER_PREFERENCE);
+
+    /* This is pretty important. We want to use the most secure TLS cipher
+    suite that we can. This simple list only allows ciphers suites which employ
+    ECDHE (Elliptic Curve Diffie-Hellman with Ephemeral keys) for encryption
+    key agreement for Perfect Forward Secrecy and we want to use AESGCM for
+    efficient authenticated encryption. */
+    if (0 == SSL_CTX_set_cipher_list(*tls_ctx, "ECDHE+AESGCM")) {
```

Comment:
First off "EECDH+AESGCM" is a great default. I was worried that older versions of OpenSSL such as those on CentOS and a few other distributions we build packages for wouldn't support it. Hence the temptation to suggest "EECDH+AESGCM:EDH+AESGCM:AES256+EECDH:AES256+EDH" or similar.

Introducing EDH ciphers comes with more trouble though as by default OpenSSL will pick a 1024-bit key for these exchanges, I wish I knew why. We'd have to take care of this by providing a stronger EDH parameter which is a hassle.

In the end it turns out that the OpenSSL version provided with CentOS doesn't even support TLSv1.2 yet, thus on those distributions we can default to building and linking our own version. This also means we can stick to "EECDH+AESGCM", which is great.

It might be nice to allow this to be configurable though, with the "EECDH+AESGCM" default. If a problem were to pop up in the future it could then be circumvented by changing the configuration instead of having to rely on us to provide a new package, or recompiling it yourself.